Added support for typed manifest immutable strings.

jocelyn opened the pull request:
For review and comments.

jocelyn committed:
Added support for typed manifest immutable strings.

manuseiffel commented on Src/C/run-time/eif_macros.h:
```diff
@@ -1152,15 +1154,21 @@ RT_LNK void eif_exit_eiffel_code(void);
+#define	RTMS(s)				makestr_with_hash(s,strlen(s),EIF_FALSE,0)
+#define	RTMS_EX(s,c)		makestr_with_hash(s,c,EIF_FALSE,0)
+#define	RTMS_EX_H(s,c,h)	makestr_with_hash(s,c,EIF_FALSE,h)
+#define RTMS_EX_O(s,c,h)	makestr_with_hash_as_old(s,c,EIF_FALSE,h)
```
```suggestion
#define	RTMS_EX_O(s,c,h)	makestr_with_hash_as_old(s,c,EIF_FALSE,h)
```

manuseiffel commented on Src/C/run-time/eif_macros.h:
```diff
@@ -1152,15 +1154,21 @@ RT_LNK void eif_exit_eiffel_code(void);
+#define RTMS_EX_O(s,c,h)	makestr_with_hash_as_old(s,c,EIF_FALSE,h)
+#define	RTMIS8_EX(s,c)		makestr_with_hash(s,c,EIF_TRUE,0)
+#define	RTMIS8_EX_H(s,c,h)	makestr_with_hash(s,c,EIF_TRUE,h)
+#define RTMIS8_EX_O(s,c,h)	makestr_with_hash_as_old(s,c,EIF_TRUE,h)
```
```suggestion
#define	RTMIS8_EX_O(s,c,h)	makestr_with_hash_as_old(s,c,EIF_TRUE,h)
```

manuseiffel commented on Src/C/run-time/eif_macros.h:
```diff
@@ -1152,15 +1154,21 @@ RT_LNK void eif_exit_eiffel_code(void);
+#define	RTMS32(s)			makestr32_with_hash(s,strlen(s),EIF_FALSE, 0)
+#define	RTMS32_EX(s,c)		makestr32_with_hash(s,c,EIF_FALSE,0)
+#define	RTMS32_EX_H(s,c,h)	makestr32_with_hash(s,c,EIF_FALSE,h)
+#define RTMS32_EX_O(s,c,h)	makestr32_with_hash_as_old(s,c,EIF_FALSE,h)
```
```suggestion
#define	RTMS32_EX_O(s,c,h)	makestr32_with_hash_as_old(s,c,EIF_FALSE,h)
```

manuseiffel commented on Src/C/run-time/eif_macros.h:
```diff
@@ -1152,15 +1154,21 @@ RT_LNK void eif_exit_eiffel_code(void);
+#define RTMS32_EX_O(s,c,h)	makestr32_with_hash_as_old(s,c,EIF_FALSE,h)
+#define	RTMIS32_EX(s,c)		makestr32_with_hash(s,c,EIF_TRUE,0)
+#define	RTMIS32_EX_H(s,c,h)	makestr32_with_hash(s,c,EIF_TRUE,h)
+#define RTMIS32_EX_O(s,c,h)	makestr32_with_hash_as_old(s,c,EIF_TRUE,h)
```
```suggestion
#define	RTMIS32_EX_O(s,c,h)	makestr32_with_hash_as_old(s,c,EIF_TRUE,h)
```

manuseiffel commented on Src/C/run-time/eif_macros.h:
```diff
@@ -1203,6 +1211,32 @@ RT_LNK void eif_exit_eiffel_code(void);
 			} \
 			r = rs; \
 		}
+#define RTCOMIS8(r,b,n,s,c,h) \
```
Might be a good idea to avoid code duplication here.

manuseiffel commented on Src/C/run-time/include/rt_interp.h:
```diff
@@ -300,7 +300,13 @@ extern "C" {
 #define BC_POSTFAIL	(unsigned char) 177
 
-#define MAX_CODE                177    /* Maximum legal byte code */
+/* Immutable Strings */
```
We have some unused entries (look for BC_NOT_USED), we should use them first before adding new ones.

jocelyn replied:
Ok I will check

manuseiffel commented on Src/C/run-time/include/rt_macros.h:
```diff
@@ -96,13 +96,31 @@ extern "C" {
 		(egc_str32set)((EIF_REFERENCE) string, u); \
 	}
 #define RT_STRING32_SET_HASH_CODE(string, hash)
+#define RT_IMMSTRING8_MAKE_FROM_STRING(obj,str) \
```
I'm not too keen about this implementation. I would prefer the same of what we had. Because in the code below when you create an immutable string it is very expensive: one normal string creation and then the immutable string creation.

jocelyn replied:
I am not a big fan neither, however it was discussed elsewhere, and we will probably update the IMMUTABLE_STRING_... interface, to add proper creating procedure from a C pointer and a given length. And probably change implementation, and rely more on the compiler, than on the runtime for supporting manifest IMMUTABLE STRINGs.

manuseiffel commented on Src/C/run-time/plug.c:
```diff
@@ -415,7 +415,25 @@ rt_public EIF_REFERENCE makestr_with_hash (const char *s, size_t len, int a_hash
 	RT_GC_WEAN(string);			/* Remove protection */
 
-	return string;
+	if (is_immut == EIF_TRUE) {
```
I'm not keen on this since we are performing two allocations for creating IMMUTABLE strings.

jocelyn replied:
agree, see previous reply

manuseiffel commented on Src/framework/parser/parser/eiffel.y:
```diff
@@ -3814,7 +3814,7 @@ Manifest_value: Boolean_constant
 	|	Nosigned_real
 			{ $$ := $1 }
-	|	Default_manifest_string
+	|	Manifest_string
```
I think we want to keep Default_manifest_string

jocelyn replied:
Given we don't want anymore
str: IMMUTABLE_STRING_32 = {IMMUTABLE_STRING_32} "abc" , 
I think I agree, I need to check a few more things.

manuseiffel commented on Src/Eiffel/eiffel/AST/visitor/ast_value_i_generator.e:
```diff
@@ -85,11 +85,13 @@ feature {NONE} -- Implementation
-			create {STRING_VALUE_I} last_value.make (a_string.value, False)
+				-- A string constant of the form:
+				--  str: STRING_8 = "abc"
+				--  str: STRING_32 = {STRING_32} "abc"
```
I do not think we should allow this. The type of the constants is telling us what to do.

jocelyn replied:
Ok, my initial idea was to allow s: READABLE_STRING_GENERAL = {IMMUTABLE_STRING_32} "abc" 
but I doubt this is really useful, and for now not implemented this way.
So I will change code, to rely on the constant type declaration, and not rely on typed manifest strings.

jocelyn committed:
Merge remote-tracking branch 'origin/master' into manifest_strings

jocelyn committed:
Support manifest string directly from generate C string, no intermediate STRING object.

jocelyn committed:
Updated to use make_from_c_byte_array for both *_STRING_32 and *_STRING_8 . Added autotest cases.

jocelyn committed:
Updated to use {ISE_RUNTIME}.new_*string_* functions.

Note: still an issue to keep emalloc_as_old behavior for once manifest strings.

jocelyn committed:
Made unicode_conversion interface more flexible.

jocelyn committed:
Merge remote-tracking branch 'origin/master' into manifest_strings

jocelyn committed:
updated no dotnet facade

jocelyn commented:
Integrated manually on subversion repository.
